### PR TITLE
avoid apply snapshot if it is the same as the current one

### PR DESCRIPTION
**Pull request by mattiamanzati**
Should address #251 
I think that checks in reconciler and applySnapshot should be enough, and they are actually pretty cheap. Am I missing something else?

### CODE REVIEW

File: src/types/type.ts [hunk 1]
```diff
@@ -89,7 +89,9 @@ export abstract class ComplexType<S, T> implements IType<S, T> {
     }
 
     reconcile(current: Node, newValue: any): Node {
-        const { parent, subpath } = current
+        if (current.snapshot === newValue)
```
Comment (mweststrate) [inline src/types/type.ts]: Q: is this check performed _before_ apply preProcessedSnapshot hooks? I think it should be before, for symmetry purposes, as the stored snapshot is the postProcessedSnapshot. Not sure about it, but there might be problems if this optimization is applied after preProcessing (although I can't think of one, maybe you can :-P).

Further LGTM :)

Comment (mattiamanzati) [inline src/types/type.ts]: A: This is always called before preProcess, as this check is performed in node "applySnapshot", and after that the type applySnapshot kicks in, so should be everything fine  :) I can't think of issues too, maybe will just end up skipping always the perf improvement ;)

**Comment by mweststrate** [comment]
Check, feel free to merge :)

Op di 18 jul. 2017 om 23:43 schreef Mattia Manzati <notifications@github.com

**Comment by mattiamanzati** [comment]
Thanks for your time! :)

**Comment by mweststrate** [comment]
Thanks for the PR 😁

Op di 18 jul. 2017 23:52 schreef Mattia Manzati <notifications@github.com>:

**Pull request merged**